### PR TITLE
#52 Refactor of setting to use new error type/handling

### PR DESCRIPTION
**Pull request by GitCodeOli**
New generic type used throughout settings, gets rid of FatalProviderError type.

FatalProviderError is replaced with a new interface called Error

```go
type Error interface {
	error
	MissingKeys() []string
	FatalError() error
}
```
This Error interface is used by the ProviderError struct which is seen throughout the Settings package. 

The aim of removing FatalProviderError was to get rid of situations where errors needed to be casted to decipher whether the error thrown was indeed fatal or "simple" e.g:

```go
	if fatalProviderError, isFatal := err.(FatalProviderError); isFatal {
		// if the error is fatal, propagate it back up
		return map[string]Setting{}, fatalProviderError
	}
```

Tests have been modified inline with the original implementation but for swapping out the new type and associated functions.

Much of the surrounding code around the error's has been left untouched but if any changes are requested I'm happy to pick up in a separate issue.

p.s. Apologies for the big PR!

### CODE REVIEW

File: aggregate_provider_test.go [hunk 1]
```diff
@@ -14,10 +14,10 @@ var dummySettingsProvider3 = dummySettingsProvider{
 	"TEST_SETTING_3": Value("false"),
 }
 
-func fatallyErroringDummySettingsProvider(keys ...string) (map[string]Setting, ProviderError) {
-	return nil, FatalProviderError{
-		SimpleProviderError: SimpleProviderError(keys),
-		Err:                 errors.New("could not load the thing to load settings"),
+func fatallyErroringDummySettingsProvider(keys ...string) (map[string]Setting, Error) {
+	return nil, ProviderError{
```
Comment (alistairjudson) [inline aggregate_provider_test.go]: Please use the constructor function here.

File: error.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// Error is a generic error wrapper that can be used to list missing keys if errors occur.
+// If a fatal error has been added to the type this should signify a type of error where execution should cease.
+type Error interface {
+	error
+	MissingKeys() []string
+	FatalError() error
+}
+
+// NewError returns a list of keys that could not be found or aren't set in the searched location.
+func NewError(keys ...string) ProviderError {
```
Comment (alistairjudson) [inline error.go]: now that we've renamed the type to ProviderError, we should call this NewProviderError see https://github.com/intelligentpos/edinburgh-tech-guide/blob/master/backend/STYLE_GUIDE.md#constructors

File: error.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	return ProviderError{
+		missingKeys: keys,
+	}
+}
+
+// NewFatalError returns a fatal error
+// and a list of keys that were supplied to the service that threw the fatal error.
+func NewFatalError(err error, keys ...string) ProviderError {
```
Comment (alistairjudson) [inline error.go]: Same as above probably should be NewProviderErrorWithFatalError or maybe NewFatalProviderError, what do you think @alessandrozucca .

File: file_loader.go [hunk 4]
```diff
@@ -14,33 +14,31 @@ type File struct {
...
-	}
+	return map[string]Setting{}, NewFatalError(err, keys...)
+
 }
 
 // load un-exported, unguarded, internal workings of Load
-func (f *File) load(keys ...string) (map[string]Setting, ProviderError) {
-	errs := SimpleProviderError{}
+func (f *File) load(keys ...string) (map[string]Setting, Error) {
+	errs := ProviderError{}
```
Comment (alistairjudson) [inline file_loader.go]: Please use the constructor here.

File: file_loader_test.go [hunk 5]
```diff
@@ -63,9 +63,9 @@ func TestProvider_Load2(t *testing.T) {
 
 func TestFatalProviderError_Error(t *testing.T) {
 	keys := []string{"here", "are", "some", "keys"}
-	fpe := FatalProviderError{
-		SimpleProviderError: SimpleProviderError(keys),
-		Err:                 errors.New("something went really wrong"),
+	fpe := ProviderError{
```
Comment (alistairjudson) [inline file_loader_test.go]: again constructor?

File: file_loader_test.go [hunk 6]
```diff
@@ -75,9 +75,9 @@ func TestFatalProviderError_Error(t *testing.T) {
 
 func TestSimpleProviderError_GetMissingKeys(t *testing.T) {
 	keys := []string{"MISSING_KEY", "MISSING_KEY_2"}
-	sp := SimpleProviderError(keys)
+	sp := ProviderError{missingKeys: keys}
```
Comment (alistairjudson) [inline file_loader_test.go]: please use the constructor.

File: os_loader.go [hunk 7]
```diff
@@ -3,20 +3,20 @@ package setting
...
 
 // OSLoad will load environment variables into settings
-// if they are found, and will return a SimpleProviderError
+// if they are found, and will return a Error
 // if a key is not found.
-func OSLoad(keys ...string) (map[string]Setting, ProviderError) {
+func OSLoad(keys ...string) (map[string]Setting, Error) {
 	hits := map[string]Setting{}
-	misses := SimpleProviderError{}
+	misses := ProviderError{}
```
Comment (alistairjudson) [inline os_loader.go]: Probably should use the constructor.

File: aggregate_provider_test.go (hunk 1 above)
Comment (GitCodeOli) [reply aggregate_provider_test.go]: Done

File: error.go (hunk 3 above)
Comment (GitCodeOli) [reply error.go]: @alessandrozucca I'll wait for response

File: error.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
+package setting
+
+import (
```
Comment (alessandrozucca) [inline error.go]: not blocking, if you have only 1 import you can just do `import "strings"`

File: error.go (hunk 3 above)
Comment (alessandrozucca) [reply error.go]: I would go for `NewFatalProviderError`

**Comment by GitCodeOli** [comment]
@alessandrozucca and @alasdairlaw updates pushed as requested. Let me know if there is anything else.

**Comment by GitCodeOli** [comment]
Closes #52

**Pull request merged**